### ISSUE TITLE
Thoughts

### ISSUE THREAD

**Issue by brandonkal** [graphile/migrate#1]
This is a very intriguing approach to database migrations! I like the idea of watching a single file during development.

I have a few questions:

1. Why Javascript?

For a fast CLI tool it would appear Go or an otherwise compiled language would be a great choice. i.e. having a single binary to deal with, working with bytes of the files directly rather than converting to strings, etc... I ask not to push one choice over the other but because I am curious if this is because of some some specific plans, i.e. "this module will eventually become a part of postgraphile".

2. What is the shadow database doing? Because current.sql is idempotent, it is not exactly clear how the shadow database is helping.

3. How would branching work out? I can imagine a lot of merge conflicts on `current.sql` and even within the migrations folder.

I realize this tool is brand new but FYI, it is not possible to commit a second migration:
```
Error: Previous migration with hash 'sha1:f2dbea7bf3b31f50e57be68cb3e6fb4c9a6b5924' doesn't match '000002.sql''s expected previous hash 000001.sql
```

**Comment by benjie** [comment]
1. The performance of the language itself is not super critical, it just needs to read data from a file and post it to PostgreSQL. It runs in node in milliseconds so Go would have no human perceptible advantage, and JS/TS is in line with the rest of the Graphile suite minimising mental context switching costs.
2. It ensures your committed migrations are consistent; it’s possible to mess up idempotency (e.g. by including changes from a different branch, or by iterating the file over time and losing the old state), but the shadow database makes sure even if you do that committed migrations can run against a clean/other persons db.
3. Committed migrations folder conflicts should never occur assuming you only commit linearly on master as recommended. current.sql may get merge conflicts if you merge two feature branches that each have pending DB migrations, like any other source file. My workflow typically doesn’t do this, as feature branches are merged to master and then the migration committed, but solving this for merging feature branches into each other is not a project goal rn.

2nd commit: noted; will fix. Commit doesn’t have sufficient tests yet, hench the “highly experimental” in the README 😉

**Comment by brandonkal** [comment]
Thank you.

1. Quite reasonable
2. So the idea with the shadow database is to maintain a "clean" (i.e. before current.sql) database dump. As far as I could tell from the code, the dump is not yet implemented. It would appear that the idea is to allow other tools to manage other parts of the schema but as far as I can predict, those outside changes would not be captured by the shadow database either. It would seem that everything in the `migrations/committed` folder would contain all the instructions needed to reach the current DB state. What intent am I missing here?

**Comment by benjie** [comment]
> So the idea with the shadow database is to maintain a "clean" (i.e. before current.sql) database dump.

It also gets reset and has the migrations re-ran to ensure the latest migration applies cleanly against a "pure" database.


> As far as I could tell from the code, the dump is not yet implemented.

Correct; though it's probably just going to fork out to `pg_dump` so shouldn't be too complex.

> It would appear that the idea is to allow other tools to manage other parts of the schema but as far as I can predict, those outside changes would not be captured by the shadow database either.

The shadow database *should* include those in case your database functions/etc have a dependency on them - that's where the `afterReset` command comes in currently. In the starter I'm working on it's used to install the `graphile_worker` database schema so that the rest of your schema can rely on a job queue without having to manage the job queue schema itself. I need to add more hooks so those schemas can be kept up to date; this is not solved yet.

> It would seem that everything in the migrations/committed folder would contain all the instructions needed to reach the current DB state. What intent am I missing here?

You're just missing the `.gmrc` -> `afterReset` / etc hooks for integrating external dependencies (e.g. graphile-worker). 

Here's the `.gmrc` from my current starter; I'm planning to release it in the next month.

```json
{
  "pgSettings": {
    "search_path": "app,app_private,app_hidden,public"
  },
  "placeholders": {
    ":DATABASE_AUTHENTICATOR": "!ENV",
    ":DATABASE_VISITOR": "!ENV"
  },
  "afterReset": ["afterReset.sql", { "command": "cd server/dist/worker && graphile-worker --once" }]
}
```

The referenced `afterReset.sql` contains:

```sql
BEGIN;
GRANT CONNECT ON DATABASE :DATABASE_NAME TO :DATABASE_OWNER;
GRANT CONNECT ON DATABASE :DATABASE_NAME TO :DATABASE_AUTHENTICATOR;
GRANT ALL ON DATABASE :DATABASE_NAME TO :DATABASE_OWNER;

-- Some extensions require superuser privileges, so we create them before migration time.
CREATE EXTENSION IF NOT EXISTS plpgsql WITH SCHEMA pg_catalog;
CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA public;
CREATE EXTENSION IF NOT EXISTS citext WITH SCHEMA public;
CREATE EXTENSION IF NOT EXISTS pgcrypto WITH SCHEMA public;
COMMIT;
```

**Comment by brandonkal** [comment]
This sounds perfect.

I've been thinking about the concept of dependencies in the DB schema. For instance, sharing schema and utilities that are not application specific.  I am sure this is out of scope right now, but just a thought...
afterReset could be extended to a dependencies section. These could be their own migrate folders, SQL files, or commands. The package.json of the DB.

**Comment by benjie** [comment]
There are people in the graphile chat who do something like this using PostgreSQL extensions (you can create your own sql-only extensions).

**Comment by benjie** [comment]
Fixed in https://github.com/graphile/migrate/commit/0f3378571d4f960f7c1259d6a47c7bcd657cf094

**Comment by benjie** [comment]
(Released as v0.0.1-alpha.1)

**Comment by brandonkal** [comment]
Yes, SQL-only extensions are an option but the problem with that is that it requires  filesystem access which is not an option for managed DBaaS or RDS.

**Comment by benjie** [comment]
`afterReset` (and related hooks) may or may not already do what you need (or be able to do it without much work). In case you're not familiar with my work, I'm a big fan of hooks and plugins... 😁

**Comment by brandonkal** [comment]
Yes, I quite like the approach here and the hook system. Though depending on how it is used it could
become complex. i.e. running another instance of graphile-migrate in a hook would probably cause a 
headache as they are not aware of each other. For what it is worth, I thought this idea out a little more.

Again, I preface this knowing it is likely out of scope here. As such, it probably makes more sense as a different tool. If you agree, I'll look into building such a thing with the DX inspired by graphile-migrate.

For instance, here is an imagining of a bookkeeping file. The DB table would match this:
```yaml
project: graphile-starter
# version is deduced from migrations list e.g. v6

# Requires only exists to verify the folders are available. All changes are listed in "migrations".
requires:
- graphile-worker@0.1.0
- users@2
# This could also be used by graphile-migrate to decide what to watch.
# e.g. these files would be watched if they exist:
# main/migrations/current.sql, users/migrations/current.sql, graphile-worker/main/current.sql

requires_dev:
- dummy_data_for_dev@1

# There is only one DB timeline here. This is important.
# The upgrade from users@1 to users@2 may contain several migrations on its own,
# but these are combined to a single transaction between 5.sql and 6.sql when applied to the DB.
migrations:
- graphile-worker@0.0.8
- users@1
- 1
- 2
- 3
- graphile-worker@0.1.0
- 4
- 5
- users@2
- 6

after_reset: ["afterReset.sql", { "command": "cd server/dist/worker && graphile-worker --once" }]
dev:
    after_migration: dummy_data_for_dev@1
```
The one thing I am not sure of is how such modular flexibility would work with the chained hash approach.
I see that graphile-migrate is using this (very git-like and a great idea!).
I'd be interested in your thoughts here.

The folder structure would be something like this:
```
db
├── dummy_data_for_dev
│   └── migrations
│       └── committed
│           └── 1.sql
├── main
│   ├── migrations
│   │   ├── committed
│   │   │   ├── 1.sql
│   │   │   ├── 2.sql
│   │   │   ├── 3.sql
│   │   │   ├── 4.sql
│   │   │   ├── 5.sql
│   │   │   └── 6.sql
│   │   └── current.sql
│   └── plan.yml
└── users
    ├── migrations
    │   └── committed
    │       ├── 1.1.sql
    │       ├── 1.2.sql
    │       ├── 1.sql
    │       └── 2.sql
    └── plan.yml
```
The workflow would be quite similar.
```
# graphile-migrate commit <package-name or main if ommitted>
graphile-migrate commit users
graphile-migrate commit
```

**Comment by benjie** [comment]
WARNING: stream of consciousness incoming - you can see my opinions changing as you read through!

This is interesting. Here's some thoughts inspired by the above:

- Associated tools (e.g. graphile-worker) may not be using the same migration system, so we can't rely on that
- Associated tools _may not_ have backwards-compatible migrations, so just running the migration in `afterReset` with the latest version may not be sufficient. This is where the plan file could be a solution; but it also mandates a plan file.
- I'm not 100% sure that my current hashing approach is the best way to go - updates to the DB (e.g. PG 9.6 -> PG 12) or unexpected data (e.g. preventing you from applying a new constraint on production but not on dev) may require previous commits to be edited. BUT I've seen people editing older commits and then resetting the database rather than following a linear migration path and that is something I wanted to avoid/prevent with graphile-migrate.
- The versioning idea in the plan is interesting

If you plan to build your own tool, it may be worthwhile to build it on `graphile-migrate` as our internals are probably compatible with what you'd be making - feel free to open pull requests exposing our individual methods, but for those PRs to be merged those interfaces will need tests to prevent regressions.

I'm currently wondering if a migration could implicitly be "run this command"; e.g. in current.sql:

```sql
--! npx graphile-worker@7.3.0 migrate
```

This way it gets tracked like any other migration, but the actual DB commands are handled externally (and with an explicit software version). You could probably call `graphile-migrate` from here too, but to do so we'd probably need a namespace argument; how about

```sql
--! graphile-migrate --path db/users/migrations --namespace users --up-to 7
```

If you'd tried to run it without the namespace it'd error because the migrations table would not match. I think it'd handle the correct ordered inserts into the migrations table too (e.g. if `users` migration 7 failed, then 1-6 would still be performed and added to the migrations table, you can then fix migration 7 and try again by running the same migration because it won't have been locked yet).

I'm not exactly sure how this would work with watch mode. Maybe something like:

```sql
--! graphile-migrate --path db/users/migrations --namespace users --up-to 7
--?WATCH! graphile-migrate --path db/users/migrations --namespace users --current
```

Maybe we need something more explicit, like:

```sql
--@require users
```

That'd tell `graphile-migrate` to watch the `users` namespace and run its changes; and then when it got turned into a final migration it'd pin the specific migration version to run up to. `graphile-migrate commit` could also detect this and run a commit in the `users` folder before running a commit in the main folder.

I'm really liking this latest approach, because it's super flexible but also has implicit database bookkeeping (rather than a bookkeeping file), and is explicit without being verbose.

Also we currently seed all files with `-- Enter migration here`; but we could make this a customisable template that users could populate with, e.g.:

```sql
-- Reference our dependencies
--@require users
--@require worker
--@require orgs
--
-- Write the migration below this line:
-------------------------------------------
```

That way all their migrations will automatically reference all the different dependencies.

**Comment by brandonkal** [comment]
Also a stream of consciousness...
These are some really interesting ideas. I do like the special comment approach.

My concern really is that this is _too flexible_. If dependencies are allowed in the migrations themselves, then you can easily run into situations where the dependency depth is too much. For instance, if a "users" module requires another dependency, then it probably doesn't make sense as its own module. Ideally, they should be able to evolve on their own.

My idea with modules is to have small starting points for distinct objects. The main project has explicit dependencies but the modules only have dependencies on interfaces. This is probably really hard to achieve in practice. This is where the bookkeeping table/file comes into play. It enables the system to track the exact order of every migration run against the database.

Here is a made up example:
- You build an enterprise app to keep track of favorite colors. The first migration looks something like this:
```sql
-- 01.sql
CREATE TABLE users (
    id serial,
    name text not null,
    fav_color text not null,
    is_admin boolean,
    pass_hash text not null
);
```
- Management asks for a blog so users can talk about their favorite colors
```
-- 02.sql
CREATE TABLE blog_posts (
    id serial,
    user_id int references users(id),
    title text not null,
    post_content text not null,
    category_id int references categories(id)
);
CREATE TABLE categories (
    id serial,
    name text not null
);
```
- Now on a second project, you know this one will need a blog as well. You copy 02.sql from the old project and make it into a reusable module: `db/blog/committed/01.sql`

Alone, this will not work as the project's first migration (it has a dependency on the users.id interface). It itself _could_ depend on a users module but really it doesn't need to. All it cares about is that a users table exists that has an id field.

So here is an idea:
Each module could have its own verify.sql file. This file exists in the root of a module file and is not allowed to modify the DB schema. If this file exists, the migration tool runs this file and if the result of the query returns false or throws, the commit is not allowed. If users wanted explicit dependencies, the verify.sql could check for that: `SELECT version FROM migrate_version WHERE module = 'dependencyA'`. Otherwise, it could check only for what it needs, e.g. a utility function or a table with a specific column.

Before I discovered postgraphile, I was looking into https://www.cannercms.com/. It really is a cool tool to help speed up backend creation. You write JSX-like schema and it generates react code, graphQL schema, and database schema from that definition. The problem I ultimately had with it was that it has too much magic. On the other hand, postgraphile feels like the correct abstraction.

I feel that having dependencies in the SQL files is a bit too much magic as information is hidden without looking through all the committed files. I do like the idea for commands that may need to be run in between migrations but it would still have to be surfaced with a new cli command.

So a few notes why I am still leaning towards using a plan:
- Provides a high level overview of what will happen on a migration. No need to look into the committed folder to see if a command will be executed. There is also the security factor. Embedded dependencies and magic commands could work if the migration tool had a command that printed out every planned step in order, including what it is going to execute in the local shell.
- Two way dependency verification, while still being portable. I did find embedding current and previous hashes in the files to be messy when working with the module concept. So this is what I settled on:

```yaml
project: Duck
migrations:
  - c: [000001__user_table.sql, users, 40b2e701b7b0b0994199646b044b82d5f03d138f]
  - c: [000001__long_name.sql, main, c484e08abb6f52a2a04b27a469a72b82763ae865]
  - c: [000002__user_pass.sql, users, e877344422fa2523673ffacbbc0fede8902fa1bb]
dependencies:
  - main
  - users
```
That last file: `000002__user_pass.sql`
```sql
-- requires: 40b2e701b7b0b0994199646b044b82d5f03d138f
-- duck migration: user pass

CREATE TABLE user_pass...
```
So you can see, embedded in the file it tracks its dependency on its _previous migration within that module_. In this way, the module is portable between different database projects. Because technically, `000001__long_name.sq` of the main module is not a requirement for the users module, but as far as the database is concerned, it still is (extra safety). The sha1 of the file is included in the plan file. In this way, the migration tool can easily reference the two dependencies.

In order for the `000002__user_pass.sql` migration to be applied, the migration tool checks two columns to be sure it is safe to run the file. Both must be already installed: `previous` which should match `c484e08abb6f52a2a04b27a469a72b82763ae865` and `40b2e701b7b0b0994199646b044b82d5f03d138f` which is included in the SQL file being executed.

Also note that I am naming migration files. These names come from the metadata, so when `commit` is run, this info is placed into the file name, just to make browsing the committed folder easier.

**Comment by benjie** [comment]
The file naming is something I'd be interested in a small PR for; take a metadata comment (if present) and `slugify(...)` it and append to the name. I decided it was too much effort to expect people to name things when committing, but we could allow specifying it on the command line (`graphile-migrate commit -m"Enable RLS throughout"`) or in the file (smart comment); this would also be helpful when performing the migrations as we can output the descriptions rather than just file names. Are you interested in contributing that?

---

One advantage of the bookkeeping approach is it allows you to "slot in" an earlier migration if you end up having issues deploying it to production - e.g. a quick `UPDATE` command to fix some invalid data in a table that won't pass the constraint added in the next commit.

One disadvantage is it's another thing for the user to manage/learn/know — one of my aims with `graphile-migrate` was to have as close to zero learning curve as possible: write migrations in the SQL you already know, and then commit them with one command and apply them with another. However I guess book keeping could be done automatically for you unless you want to handle it manually.

I'm not sure how watch mode (particularly dependencies) would work with multiple modules; would you make it so you only watch one module at a time?

How would a user add the dependency, is that manual or somehow automated?

How are the modules distributed?

**Comment by brandonkal** [comment]
I am currently overwhelmed with a long running project but feel free to ping me for that PR if it still needs doing in a couple of months. I prefer the smart comment approach because it means the `graphile-migrate commit` can be part of the CI.

---
It is nice to be able to "slot" something in if required, but I agree the file shouldn't normally be manually edited.

In the simple developer experience of `graphile-migrate` the way I see modules working is:

1. `$ graphile-migrate watch`
2. The tool looks at the dependencies key in the bookkeeping file and automatically starts watching `db/{depName}/current.sql` for each key. `main` is a default "dependency" so most work would happen at `db/main/current.sql`
3. Developer works on any of the current.sql files live.
4. To commit a module it is just `graphile-migrate commit -m users` for a dependency otherwise "main" is the default for the "-m" flag. So usually just `graphile-migrate commit`

To add (create) a module manually: create the folder containing a current.sql and add that folder name to the dependencies key in the bookkeeping file. If we can assume that `graphile-migrate` owns the db folder, that last step isn't really necessary as it could just look for folders in its search path.
Not too hard, but ideally a little command would initialize things.

To add a module with commits it would be the same. But this would introduce another command e.g. `graphile-migrate import <module>` that would read all SQL files in the new module's committed folder and add those migrations to the bookkeeping file.

As far as distribution, curl or git submodules are options. Though I would prefer the former to not introduce complexity and just commit the module files alongside the rest of the migrations. I see it mostly as a starting templates that you can compose and then iterate on their own. Checking for updates for DB modules from a central repository wouldn't really be an option with this simple approach but that is good. Otherwise you have a distributed monolith where they can't evolve independently (well you _could_ if the change was made to main, but then it is coupled).

---

The bookkeeping file would be the same as the `.gmrc`
But to brainstorm some ways to keep that file simple:
dependencies and dev_dependencies are instead just folders. Perhaps even just a `modules` folder and a `dev` folder. Each folder is then automatically picked up.
The same goes for afterReset. Instead, the tool just looks for a `db/hooks/afterReset.sql` file. This file can contain magic shell comments for other tools.

The sha of each file doesn't really need to be stored as that can be found by reading all files into memory and calculating their hash. The safety really comes from comparing the filesystem with the hashes already applied in the DB. The tool can decide if it needs to cache the list of hashes to a gitignored file. It shouldn't be much of a problem at all until hundreds of migrations exist. The files still store their one requires hash.

The only thing I am not sure of is how you could eliminate the bookkeeping file and still persist the order of commits: `users1, users2, main1, main2, users3`

That order cannot be guaranteed. Although ideally the users3 migration doesn't change anything from main2, it is still possible that some discrepencies would arise without bookkeeping. But perhaps this is where the pg_dump comes in?

---

So where does the pg_dump come in? Is the idea to commit this to git as well? Does it always hang one commit behind the development database? I am still trying to wrap my head around this.
From the readme:

> 
> 1. reset the shadow database to the latest dump
> 2. apply the current migration to the shadow database, and replace the dump
> 3. move the current migration to committed migrations (adding a hash to prevent tampering)

1. I understand why this is necessary. But under normal operation, what would ever cause the shadow database to drift?

If the dump doesn't yet exist (.gitignored), I assume it would reset completely and run ALL migrations (as the code is currently implemented)

2. So this is the first time `current.sql` is being executed against the shadow rather than the development DB correct?

The way I see it now:
1. On every commit, it resets to zero, runs ALL migrations, including the one about to be committed, and runs pg_dump.

Is this the intended workflow now? (I am assuming eventually this would be wrapped in to a setup command)
```
$ createdb dev
$ createdb shadow
# Create and edit config files Then:
$ psql dev < the_shadow_db_dump.sql
$ graphile-migrate commit
$ graphile-migrate migrate
$ graphile-migrate watch
```
Then under normal circumstances, you never run `migrate`  to the development database (unless merging from another dev) because every commit was already executed before commit (as current.sql). I see `watch` wraps `migrate`, but this is mostly a safety thing to avoid drift.

So if I understand things,
Shadow DB receives migrations on commit and Dev DB receives migrations on the start of watch.
Finally only the `commit` command reads or writes to the shadow DB.

Last question:
In the readme:

> Add graphile-migrate `check` command: reset the shadow database to the latest dump, apply the current migration to the shadow database, and output a SQL schema diff you can use to ensure no accidental changes have been made.

Is this schema diff across databases (shadow and dev) which should normally not differ unless something is missing from `current.sql` that was applied from a previous iteration or is it just a diff between shadow-pre-migration and shadow-after-current-migration?

**Comment by brandonkal** [comment]
Oh. Is there any particular reason for using the `:PLACEHOLDER_NAME` syntax over the ubiquitous `{{placeholder}}` sytax aka [Jinja/Python](http://jinja.pocoo.org/), [Mustache/Ruby/JS](https://github.com/Mustache/Mustache), or [Go](https://gowebexamples.com/templates/) which is also quite similar to JS template literals which wouldn't work in SQL `${placeholder}`.

**Comment by benjie** [comment]
> Is this schema diff across databases (shadow and dev) ... or is it just a diff between shadow-pre-migration and shadow-after-current-migration

Both, I think. Or more specifically:

- reset shadow DB to blank slate, run all the migrations
- dump shadow DB
- compare shadow DB to committed dump - they should match because the dump is from the shadow DB and the shadow DB is just the result of running all the migrations. This would reveal any inconsistencies in the system. This should *never* differ unless you've edited history.
- compare shadow DB to dev DB - the diff should only contain changes that are deliberate (e.g. things you're doing in current.sql); further differences (e.g. mis-ordered columns) may cause you development issues so this is a handy command to check that your DB doesn't need resetting.

> Is there any particular reason for using the `:PLACEHOLDER_NAME` syntax

Yes; because it's the same syntax `psql` uses, see [SQL interpolation](https://www.postgresql.org/docs/current/app-psql.html#APP-PSQL-INTERPOLATION); basically you should be able to run `current.sql` through `psql` if you want.

> If templates are allowed, the substitution result should be baked in on commit.

That explicitly should *not* happen; if you want them baked in then don't use substitutions. The substitutions enable you to do things like using different database name/role name/etc in different environments; e.g. you might want to run a few different "preview" servers for different pull requests, but running that many PostgreSQL servers is unnecessary - just run one and use different variables.

---

Thanks for discussing this with me; it's been an interesting conversation and has given me a lot of good ideas. I don't think the module system as you've described it is something that I'd want in `graphile-migrate`, but I think you could build a pretty convincing mock of it by building a small script that downloads/checks the versions from a remote URL/local folder and applies them locally using the existing `graphile-migrate` migration system. I think a linear history in this regard should be no issue.

**Comment by brandonkal** [comment]
After some more thought, it should be possible to completely avoid the bookkeeping file.
It would add a bit of complexity to the tool but would hopefully simplify the developer experience...

The schema history table would track four hash columns:
1. migration file hash
2. previous migration hash (should be the row directly above)
3. previous requires hash (last migration for that particular module
4. migration file content hash after metadata block

The file when committed would look something like this:
```sql
-- graphile migration: add users table
-- requires: hash
-- previous: hash

CREATE TABLE...
```

The tool would then have to read all the first three lines of all migrations into memory to compare with the DB. The first hash is just the hash of the entire file. This is important so the integrity of the metadata isn't changed.

The previous hash can actually change. When you import a module, **all of the previous hashes are updated to be correct**. 

So the previous hash stored in `users/committed/01.sql` may be the hash of `main/committed/01.sql` for this particular project but something different in another project. This is why the fourth hash must be recorded in the database. It isn't used for safety checks but is still recorded so the user can answer questions like "is this the same `users/committed/01.sql` in my other project?" or perhaps eventually, checking for updates. That would otherwise be impossible since we have modified the file so that the previous hash is correct for this project. Again, the hash of the contents without the metadata block can easily be calculated from the file itself, so it isn't stored explicitly in the file.

Then to solve the issue of too much magic, a "show plan" command would need to exist that would print out the order that the next migrations would occur in.

I am quite happy with this approach.

**Comment by brandonkal** [comment]
Yes, Thank you for the insightful discussion @benjie! Sorry for that big block comment. If you have a minute would you shed some light on this? I have extracted from above:

a. Where does the pg_dump come in?

b. Is the idea to commit this to git as well?

c. Does it always hang one commit behind the development database?

From the readme:

> 1. reset the shadow database to the latest dump
> 2. apply the current migration to the shadow database, and replace the dump
> 3. move the current migration to committed migrations (adding a hash to prevent tampering)

1. I understand why this is necessary. But under normal operation, what would cause the shadow database to drift?

2. This is the first time `current.sql` is being executed against the shadow rather than the development DB correct?
3. Is this correct?
Shadow DB receives migrations on `commit` and Dev DB receives migrations on the start of `watch`.
Finally only the `commit` command reads or writes to the shadow DB.

**Comment by benjie** [comment]
> a. Where does the pg_dump come in?  
> b. Is the idea to commit this to git as well?


After 'graphile-worker commit' a dump will be made and stored to disk. This should be committed. You can then review the diff to ensure that no breaking changes have take place/you've not accidentally dropped things you didn't mean to.

> c. Does it always hang one commit behind the development database?

You can run `graphile-migrate dump [--shadow]` manually to dump your development (or shadow) database; you can then `git diff` to ensure you're changing what you think you're changing. It's also useful for cleaning up changes from other branches you may have been working on.

>  But under normal operation, what would cause the shadow database to drift?

Most likely this will be caused by PG version upgrades.

> This is the first time current.sql is being executed against the shadow rather than the development DB correct?

`current.sql` is ran against the shadow database during `commit`. I can't remember if it gets put into `committed` first or not, that's an implementation detail really.

> Shadow DB receives migrations on `commit` and Dev DB receives migrations on the start of `watch`.

Correct; you can also run various commands manually to apply migrations to dev or shadow.

> only the `commit` command reads or writes to the shadow DB.

Whilst this may be true currently, it likely won't be true going forward. The shadow database is something we can reset on a whim. (The `reset --shadow` command also resets the shadow db.)
